syntax = "proto3";

package nativedroid_server;
option java_package = "org.argus.nativedroid.server";

import "java_signatures.proto";

message GenSummaryRequest {
    string apk_digest = 1;
    string component_name = 2;
    int32 depth = 3;
    string so_digest = 4;
    string jni_func = 5;
    jawa_core.MethodSignature method_signature = 6;
}

message GenSummaryResponse {
    string taint = 1;
    string summary = 2;
    int64 analyzed_instructions = 3;
}

message HasSymbolRequest {
    string so_digest = 1;
    string symbol = 2;
}

message HasSymbolResponse {
    bool has_symbol = 1;
}

message AnalyseNativeActivityRequest {
    string apk_digest = 1;
    string component_name = 2;
    string so_digest = 3;
    string custom_entry = 4;
}

message AnalyseNativeActivityResponse {
    int64 total_instructions = 1;
}

message LoadBinaryRequest {
    bytes buffer = 1;
}

message LoadBinaryResponse {
    string so_digest = 1;
    int32 length = 2;
}

service NativeDroid {
    rpc GenSummary(GenSummaryRequest) returns (GenSummaryResponse);
    rpc HasSymbol(HasSymbolRequest) returns (HasSymbolResponse);
    rpc AnalyseNativeActivity(AnalyseNativeActivityRequest) returns (AnalyseNativeActivityResponse);
    rpc LoadBinary(stream LoadBinaryRequest) returns (LoadBinaryResponse);
}